### ISSUE TITLE
Wrong lemma goes through

### ISSUE THREAD

**Issue by jad-hamza** [epfl-lara/stainless#497]
```scala
import stainless.lang._
import stainless.collection._

object NextPointer {
  def pointsToNext(m: BigInt => BigInt, l: List[BigInt], i: BigInt) = {
    0 <= i &&
    i < l.length &&
    m(l(i)) == l((i + 1) % l.length)
  }

  def badLemma(m: BigInt => BigInt, l: List[BigInt], b: BigInt): Boolean = {
    require(pointsToNext(m, l, l.length-1))

    pointsToNext(x => b, l, l.length-1)
  } holds

  def theorem() = {
    val m1: BigInt => BigInt = x => 1
    val m2: BigInt => BigInt = x => 2
    val l = List[BigInt](1)
    assert(pointsToNext(m1, l, 0))
    assert(!pointsToNext(m2, l, 0))
    badLemma(m1, l, 2)
    assert(pointsToNext(m2, l, 0))
    assert(false)
  }
}
```

`badLemma` goes through while it shouldn't, and `theorem` uses it on `m1, l, 2` to show a contradiction. The VC generated by Stainless for `badLemma` looks normal, so the problem may come from Inox?

```scala
¬pointsToNext(m, l, length(l) - 1) || pointsToNext((x: BigInt) => b, l, length(l) - 1)
```

**Comment by jad-hamza** [comment]
I was able to simplify the example, but this seems to be a Z3 bug as CVC4 does report the wrong verification conditions. We can keep this issue open to keep track of the progress.

```scala
import stainless.lang._

object SelfModulus {
  def greaterThanZero(n: BigInt) = n > 0 && n % n == 0

  def greaterThanZeroImpliesFalse(n: BigInt) = {
    require(greaterThanZero(n))
    false
  } holds

  def theorem() = {
    greaterThanZeroImpliesFalse(1)
    assert(false)
  }
}
```

**Comment by jad-hamza** [comment]
It all works well on Z3 4.8.4, the bug is there only on nativez3 and Z3 4.7.1.

**Comment by mario-bucev** [comment]
This issue is resolved, see #1260 :)
